File: prev-2024/523.continuous-subarray-sum.js
/*
 * @lc app=leetcode id=523 lang=javascript
 *
 * [523] Continuous Subarray Sum
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {boolean}
 */
var checkSubarraySum = function (nums, k) {
  const n = nums.length;
  if (n < 2) {
    return false;
  }

  let currentSum = 0;

  /** @type {Map<number, number>} */
  const remainderToLeftMostIndex = new Map();

  for (let i = 0; i <= n; i++) {
    if (i > 0) {
      currentSum += nums[i - 1];
    }
    const remainder = currentSum % k;
    if (remainderToLeftMostIndex.has(remainder)) {
      const leftMostIndex = remainderToLeftMostIndex.get(remainder);
      if (i - leftMostIndex >= 2) {
        return true;
      }
    } else {
      remainderToLeftMostIndex.set(remainder, i);
    }
  }

  return false;
};
// @lc code=end

File: prev-2024/523.continuous-subarray-sum.cpp
/*
 * @lc app=leetcode id=523 lang=cpp
 *
 * [523] Continuous Subarray Sum
 */
#include <unordered_map>
#include <vector>

using std::unordered_map;
using std::vector;

// @lc code=start
class Solution {
 public:
  bool checkSubarraySum(vector<int>& nums, int k) {
    const auto N = nums.size();
    int prefixSumMod = 0;
    unordered_map<int, int> prefixSumModToIndex;
    // Initialize to handle the case
    // where the subarray starts from the beginning
    prefixSumModToIndex[0] = -1;

    for (int i = 0; i < N; i++) {
      prefixSumMod = (prefixSumMod + nums[i]) % k;
      if (prefixSumModToIndex.count(prefixSumMod)) {
        const int index = prefixSumModToIndex[prefixSumMod];
        if (i - index > 1) {
          return true;
        }
      } else {
        prefixSumModToIndex[prefixSumMod] = i;
      }
    }

    return false;
  }
};
// @lc code=end

int main() {
  Solution sol;
  vector<int> nums = {5, 0, 0, 0};
  sol.checkSubarraySum(nums, 3);
}
